File: nxflix-agents-python/src/nxflix_agents/routers/vocabulary.py
"""Vocabulary API routes."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from nxflix_agents.models import VocabularyItem, VocabularyGenerateRequest
from nxflix_agents.models.vocabulary import PartOfSpeech
from nxflix_agents.services import VocabularyService
from nxflix_agents.providers.llm import LLMProvider

router = APIRouter(prefix="/api/vocabulary", tags=["vocabulary"])

# Singleton vocabulary service
_vocabulary_service = VocabularyService()
_llm = LLMProvider()


class VocabularyListResponse(BaseModel):
    """Response for listing vocabulary."""

    vocabulary: list[VocabularyItem]
    count: int


class VocabularySingleResponse(BaseModel):
    """Response for a single vocabulary item."""

    vocabulary: VocabularyItem


class VocabularyGeneratedResponse(BaseModel):
    """Response schema for generated vocabulary."""

    vocabulary: list[VocabularyItem]


@router.get("/", response_model=VocabularyListResponse)
async def list_vocabulary():
    """List all vocabulary items."""
    vocabulary = _vocabulary_service.get_all_vocabulary_items()
    return VocabularyListResponse(vocabulary=vocabulary, count=len(vocabulary))


@router.get("/search", response_model=VocabularyListResponse)
async def search_vocabulary(query: str):
    """Search vocabulary by word, reading, or meanings."""
    vocabulary = _vocabulary_service.search_vocabulary(query)
    return VocabularyListResponse(vocabulary=vocabulary, count=len(vocabulary))


@router.get("/by-word/{word}", response_model=VocabularySingleResponse)
async def get_vocabulary_by_word(word: str):
    """Get vocabulary by its word."""
    vocabulary = _vocabulary_service.get_vocabulary_by_word(word)
    if not vocabulary:
        raise HTTPException(status_code=404, detail="Vocabulary not found")
    return VocabularySingleResponse(vocabulary=vocabulary)


@router.get("/by-pos/{pos}", response_model=VocabularyListResponse)
async def get_vocabulary_by_pos(pos: str):
    """Get vocabulary items by part of speech."""
    try:
        part_of_speech = PartOfSpeech(pos)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid part of speech")

    vocabulary = _vocabulary_service.get_vocabulary_by_part_of_speech(part_of_speech)
    return VocabularyListResponse(vocabulary=vocabulary, count=len(vocabulary))


@router.get("/{vocab_id}", response_model=VocabularySingleResponse)
async def get_vocabulary(vocab_id: str):
    """Get a single vocabulary item by ID."""
    vocabulary = _vocabulary_service.get_vocabulary_item(vocab_id)
    if not vocabulary:
        raise HTTPException(status_code=404, detail="Vocabulary not found")
    return VocabularySingleResponse(vocabulary=vocabulary)


@router.post("/generate", response_model=VocabularyListResponse)
async def generate_vocabulary(request: VocabularyGenerateRequest):
    """AI-generate a set of vocabulary items."""
    prompt = _build_vocabulary_generation_prompt(request)

    try:
        generated = await _llm.complete_json(
            messages=[{"role": "user", "content": prompt}],
            response_model=VocabularyGeneratedResponse,
        )

        # Add generated vocabulary to service
        for vocab in generated.vocabulary:
            _vocabulary_service.add_vocabulary_item(vocab)

        return VocabularyListResponse(
            vocabulary=generated.vocabulary, count=len(generated.vocabulary)
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


def _build_vocabulary_generation_prompt(request: VocabularyGenerateRequest) -> str:
    """Build the prompt for vocabulary generation."""
    parts: list[str] = [
        "Generate JLPT N1 vocabulary data in JSON format.",
        "",
    ]

    if request.topic:
        parts.append(
            f"Generate {request.count} vocabulary items related to the topic: {request.topic}"
        )
    else:
        parts.append(f"Generate {request.count} random JLPT N1 vocabulary items.")

    if request.part_of_speech:
        parts.append(f"Focus on {request.part_of_speech.value} words.")

    parts.extend([
        "",
        "For each vocabulary item, provide:",
        "- id: unique identifier (format: vocab-XXX)",
        "- word: the word in kanji (if applicable)",
        "- reading: the reading in hiragana",
        "- meanings: array of English meanings",
        "- part_of_speech: one of: noun, verb, adjective_i, adjective_na, adverb, particle, conjunction, expression",
    ])

    if request.include_examples:
        parts.append(
            "- examples: array of 1-2 example sentences with sentence and translation"
        )

    parts.extend([
        "- synonyms: array of synonymous words (optional)",
        "- antonyms: array of antonymous words (optional)",
        "- notes: any additional notes about usage (optional)",
        "",
        'Return a JSON object with a "vocabulary" array containing the vocabulary objects.',
    ])

    return "\n".join(parts)


# Export the service for state management
def get_vocabulary_service() -> VocabularyService:
    """Get the vocabulary service singleton."""
    return _vocabulary_service

File: nxflix-agents-ts/src/routers/vocabulary.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { VocabularyService } from '../services/vocabulary.js';
import { VocabularyItem, VocabularyGenerateRequest, PartOfSpeech } from '../models/vocabulary.js';
import { LLMProvider } from '../providers/llm.js';

const llm = new LLMProvider();
const vocabularyRouter = Router();

// Singleton vocabulary service
const vocabularyService = new VocabularyService();

// Request schemas
const VocabularySearchSchema = z.object({
  query: z.string().min(1),
});

const VocabularyGenerateSchema = VocabularyGenerateRequest;

// Response types
interface VocabularyListResponse {
  vocabulary: VocabularyItem[];
  count: number;
}

interface VocabularySingleResponse {
  vocabulary: VocabularyItem;
}

// GET /api/vocabulary - List all vocabulary items
vocabularyRouter.get('/', (_req: Request, res: Response) => {
  const vocabulary = vocabularyService.getAllVocabularyItems();
  res.json({ vocabulary, count: vocabulary.length } as VocabularyListResponse);
});

// GET /api/vocabulary/search - Search vocabulary
vocabularyRouter.get('/search', (req: Request, res: Response) => {
  try {
    const { query } = VocabularySearchSchema.parse(req.query);
    const vocabulary = vocabularyService.searchVocabulary(query);
    res.json({ vocabulary, count: vocabulary.length } as VocabularyListResponse);
  } catch (error) {
    console.error('Error searching vocabulary:', error);
    res.status(400).json({ error: String(error) });
  }
});

// GET /api/vocabulary/by-word/:word - Get vocabulary by word
vocabularyRouter.get('/by-word/:word', (req: Request<{ word: string }>, res: Response) => {
  const vocabulary = vocabularyService.getVocabularyByWord(req.params.word);
  if (!vocabulary) {
    res.status(404).json({ error: 'Vocabulary not found' });
    return;
  }
  res.json({ vocabulary } as VocabularySingleResponse);
});

// GET /api/vocabulary/by-pos/:pos - Get vocabulary by part of speech
vocabularyRouter.get('/by-pos/:pos', (req: Request<{ pos: string }>, res: Response) => {
  try {
    const partOfSpeech = PartOfSpeech.parse(req.params.pos);
    const vocabulary = vocabularyService.getVocabularyByPartOfSpeech(partOfSpeech);
    res.json({ vocabulary, count: vocabulary.length } as VocabularyListResponse);
  } catch (error) {
    res.status(400).json({ error: 'Invalid part of speech' });
  }
});

// GET /api/vocabulary/:id - Get single vocabulary by ID
vocabularyRouter.get('/:id', (req: Request<{ id: string }>, res: Response) => {
  const vocabulary = vocabularyService.getVocabularyItem(req.params.id);
  if (!vocabulary) {
    res.status(404).json({ error: 'Vocabulary not found' });
    return;
  }
  res.json({ vocabulary } as VocabularySingleResponse);
});

// POST /api/vocabulary/generate - AI-generate vocabulary set
vocabularyRouter.post('/generate', async (req: Request, res: Response) => {
  try {
    const request = VocabularyGenerateSchema.parse(req.body);

    const prompt = buildVocabularyGenerationPrompt(request);

    const generatedVocabulary = await llm.completeJson<{ vocabulary: VocabularyItem[] }>(
      [{ role: 'user', content: prompt }],
      z.object({
        vocabulary: z.array(VocabularyItem),
      })
    );

    // Add generated vocabulary to service
    for (const vocab of generatedVocabulary.vocabulary) {
      vocabularyService.addVocabularyItem(vocab);
    }

    res.json({
      vocabulary: generatedVocabulary.vocabulary,
      count: generatedVocabulary.vocabulary.length,
    } as VocabularyListResponse);
  } catch (error) {
    console.error('Error generating vocabulary:', error);
    res.status(500).json({ error: String(error) });
  }
});

function buildVocabularyGenerationPrompt(request: z.infer<typeof VocabularyGenerateSchema>): string {
  const parts: string[] = [
    'Generate JLPT N1 vocabulary data in JSON format.',
    '',
  ];

  if (request.topic) {
    parts.push(`Generate ${request.count} vocabulary items related to the topic: ${request.topic}`);
  } else {
    parts.push(`Generate ${request.count} random JLPT N1 vocabulary items.`);
  }

  if (request.partOfSpeech) {
    parts.push(`Focus on ${request.partOfSpeech} words.`);
  }

  parts.push('');
  parts.push('For each vocabulary item, provide:');
  parts.push('- id: unique identifier (format: vocab-XXX)');
  parts.push('- word: the word in kanji (if applicable)');
  parts.push('- reading: the reading in hiragana');
  parts.push('- meanings: array of English meanings');
  parts.push('- partOfSpeech: one of: noun, verb, adjective_i, adjective_na, adverb, particle, conjunction, expression');

  if (request.includeExamples) {
    parts.push('- examples: array of 1-2 example sentences with sentence and translation');
  }

  parts.push('- synonyms: array of synonymous words (optional)');
  parts.push('- antonyms: array of antonymous words (optional)');
  parts.push('- notes: any additional notes about usage (optional)');
  parts.push('');
  parts.push('Return a JSON object with a "vocabulary" array containing the vocabulary objects.');

  return parts.join('\n');
}

// Export the service for state management
export { vocabularyRouter, vocabularyService };
